File: Matlab Code/Classes/RegimeManager.m
classdef (Sealed = true) RegimeManager < handle
   
    % This class is responsible for maintaining the list of regimes in
    % Imagine. It should listen to the cropManager for relevant crop changes
    % and update notify the ImagineWindowManager via events when a change
    % to the screen is required.
    
    properties
       listeners = event.listener.empty(1, 0);
    end
    
    % Implement a singleton class.
    methods (Access = private)
        function obj = RegimeManager()
        end        
        
        function regimeManagerConstructor(obj)
            cropMgr = CropManager.getInstance();
            obj.listeners(end + 1) = addlistener(cropMgr, 'CropRemoved', @obj.cropWasRemoved);             
            obj.listeners(end + 1) = addlistener(cropMgr, 'CropNameChanged', @obj.cropNameChanged);             
        end
    end
    
    methods (Static)
        function singleObj = getInstance(loadedObj)
            persistent localObj
            
            % If a RegimeManager is passed in and is not the localObj,
            % then set it as the localObj.
            if nargin >= 1
                if isa(loadedObj, 'RegimeManager') && localObj ~= loadedObj
                    localObj = loadedObj;
                    localObj.regimeManagerConstructor;
                else
                    disp('Tried passing an object that''s not a RegimeManager to RegimeManager.getInstance.');
                    % Or possibly localObj ~= loadedObj because
                    % getInstance() has already been called before
                    % getInstance(loadedObj) has had a chance because some
                    % other getInstance (eg CropManager) calls getInstance
                    % of the RegimeManager.
                    % If this is the case then we should implement a
                    % listener for 'finishedLoading' which is triggered at
                    % the end of the ImagineObject load function and
                    % triggers the work that should be done after the
                    % objects are loaded - the kinds of things that might
                    % call RegimeManager.getInstance() such as redoing the
                    % listeners.
                    % So instead of calling regimeManagerConstructor in
                    % here, we set a listener on the ImagineObject and set
                    % the callback to do the constructing work.
                end
            end
            
%             disp(['RegimeManager is empty: ',  num2str(isempty(localObj))]);
%             if ~isempty(localObj)
%                 disp(['RegimeManager is valid: ',  num2str(isvalid(localObj))]);
%             end
            if isempty(localObj) || ~isvalid(localObj)
                localObj = RegimeManager;
                localObj.regimeManagerConstructor;
            end
            
            singleObj = localObj;
            
        end
    end
    
    properties
       
        regimes = Regime.empty(1, 0);
        
    end
    
    events
       % Many of these events should pass a new regime definitions list in the eventdata.
        
       RegimeAdded % Means we should add a regime to the list.
       RegimeRemoved % Will mean the name should be removed in the window.
              
       % Means that an edit has occured. 
       % This one might just contain a structure that 
       % has relevant editted fields... (Changed to what? From what?)
       RegimeEditted
       
       % In handler for RegimeEditted we may trigger these three events.
       RegimeCategoryChanged  % Will affect any regimes the crop was in.
       RegimeColourChanged % Will mean the ImagineWindow should use different colour.
       RegimeNameChanged % Will mean the regime list should be updated.
    end
    
    properties (Dependent)
        
        % This is a cut-down version of the regimes list. It ony gives
        % those things common to regimes, like the regimeLable, the years,
        % the type, the category, etc.
       regimeDefinitions 
    end
    
    methods (Static)
       
        % Register all the different regime delegate types in this function.
        % They should have a regimeType, and a delegateClass
        function RTs = getRegimeTypes
            
            RT.regimeCategory = 'Annual';
            RT.regimeType = 'primary';
            RT.delegateClass = 'AnnualRegimeDelegate';
            
            RTs(1) = RT;
            
            RT.regimeCategory = 'Belt';
            RT.regimeType = 'secondary';
            RT.delegateClass = 'BeltRegimeDelegate';
            
            RTs(2) = RT;
            
        end
        
        function loaded = loadobj(obj)
            disp('loading regime manager');
            obj.regimeManagerConstructor;
            loaded = obj;
        end
      
        
    end
    
    methods

        
        
        
        % Adds a new regime of the given type
        function addRegime(regimeMgr, regimeCat)            
            
            RTs = RegimeManager.getRegimeTypes;
            RT = RTs(strcmp({RTs.regimeCategory}, regimeCat));
            if isempty(RT)
                error('Tried to add a regime in non-exsitent regime category.'); 
            end
                
            
            
            r = Regime(RT.delegateClass);
            
            r = r.setupRegime;
            
            if Regime.isValid(r)
               % If adding the regime was successful, add it to the list of
               % regimes.
               % We also need to notify other Managers that we've added the
               % regime.
               regimeMgr.regimes(end+1) = r;
               regimeMgr.sortRegimes;
              
               disp('Added regime');
               
                % Broadcast the event
                evtData = RegimeListChangedEventData(regimeMgr.regimeDefinitions, '', r.regimeLabel);
                notify(regimeMgr, 'RegimeAdded', evtData);

               disp('Notified: Added regime');
                
            else
                % Must have cancelled or failed. Either way, we certainly dont update
                % the regime.
            %    error('The regime''s delegate returned an object that the Regime class deemed invalid.');
            end
            
        end
        
        % Removes regime with name r, or if r is a number, at index r.
        function removeRegime(regimeMgr, r, forceTF)
            
            if (nargin < 3)
                forceTF = false;
            end
            
            % Get the regime based on the index or string in r.
            if isnumeric(r)
                if r >= 1 && r <= length(regimeMgr.regimes)
                   regimeIndex = r;
                else
                    error('Regime index passed is out of bounds.');
                end
            elseif ischar(r)
                regimeIndex = find(strcmp({regimeMgr.regimes.regimeLabel}, r), 1, 'first'); 
                if isempty(regimeIndex)
                    error('Cannot find regime with matching regimeLabel to edit.');
                end
            else
                error('Argument passed to editRegime must be a string with a valid regimeLabel of the index of the regime in the regimeList.');
            end
            
            r = regimeMgr.regimes(regimeIndex);
            
            if (forceTF)
                buttons = 'Yes';
            else
                buttons = questdlg(['Are you sure you want to remove the regime with label: ', r.regimeLabel], 'Remove Regime Check');
            end
            
            if strcmp(buttons, 'Yes') 
                rLabel = r.regimeLabel;
                delete(r);
                regimeMgr.regimes = [regimeMgr.regimes(1:regimeIndex - 1), regimeMgr.regimes(regimeIndex + 1: end)];
                                
                % Broadcast the event
                evtData = RegimeListChangedEventData(regimeMgr.regimeDefinitions, rLabel, '');
                notify(regimeMgr, 'RegimeRemoved', evtData);
                
            end
        end
        
        % Edits the regime with name r, or if r is a number, at index r.
        function editRegime(regimeMgr, r)
            
            % Get the regime based on the index or string in r.
            if isnumeric(r)
                if r >= 1 && r <= length(regimeMgr.regimes)
                   regimeIndex = r;
                else
                    error('Regime index passed is out of bounds.');
                end
            elseif ischar(r)
                regimeIndex = find({regimeMgr.regimes.regimeLabel} == r, 1); 
                if isempty(regimeIndex)
                    error('Cannot find regime with matching regimeLabel to edit.');
                end
            else
                error('Argument passed to editRegime must be a string with a valid regimeLabel of the index of the regime in the regimeList.');
            end
            
            r = regimeMgr.regimes(regimeIndex);
            rPrevLabel = r.regimeLabel;
            r = r.setupRegime;
                        
            if Regime.isValid(r)
                regimeMgr.replaceRegime(rPrevLabel, r)
            elseif isempty(r)
                % Must have cancelled. Either way, we certainly dont update
                % the regime.
            else
                error('The regime''s delegate returned an object that the Regime class deemed invalid.');
            end
                        
        end  % end editRegime
        
        % Replaces the regime with name regimeName with newRegime
        function replaceRegime(regimeMgr, regimeLabel, newRegime)

            regimeIndex = find(ismember({regimeMgr.regimes.regimeLabel}, regimeLabel), 1); 
            if isempty(regimeIndex)
                error('Cannot find regime with matching regimeLabel to edit.');
            end
            
            if Regime.isValid(newRegime)
               regimeMgr.regimes(regimeIndex) = newRegime;
               regimeMgr.sortRegimes;
               
                % Broadcast the event
                evtData = RegimeListChangedEventData(regimeMgr.regimeDefinitions, regimeLabel, newRegime.regimeLabel);
                notify(regimeMgr, 'RegimeEditted', evtData);
            end
            
        end
                
        % Constructs an array of regime defintiions.
        function regdefs = get.regimeDefinitions(regMgr)
        
            labels = {regMgr.regimes.regimeLabel};
            categoryNames = {regMgr.regimes.regimeCategory};
            types = {regMgr.regimes.type};
            startYears = {regMgr.regimes.startYear};
            finalYears = {regMgr.regimes.finalYear};
            timelineColours = {regMgr.regimes.timelineColour};
            cropNameLists = {regMgr.regimes.cropNameList};
            
            regdefs = struct('regimeLabel', labels, 'categoryName', categoryNames, 'type', types, 'startYear', startYears, ...
                        'finalYear', finalYears, 'timelineColour', timelineColours, 'cropNameList', cropNameLists);
        end
        
        function regdef = getRegimeDefinition(regMgr, regimeLabel)           
            ix = find(strcmp({regMgr.regimes.regimeLabel}, regimeLabel), 1, 'first');
            if isempty(ix)
                regdef = [];
                return
            end
            reg = regMgr.regimes(ix);
        
            labels = {reg.regimeLabel};
            categoryNames = {reg.regimeCategory};
            types = {reg.type};
            startYears = {reg.startYear};
            finalYears = {reg.finalYear};
            timelineColours = {reg.timelineColour};
            cropNameLists = {reg.cropNameList};
            
            regdef = struct('regimeLabel', labels, 'categoryName', categoryNames, 'type', types, 'startYear', startYears, ...
                        'finalYear', finalYears, 'timelineColour', timelineColours, 'cropNameList', cropNameLists);
            
        end
        
        % Provides access to external objects to query the name of a crop
        % planted in a particular regime in a particular year. Note that a
        % cell array is returned, as there may be more than one crop
        % planted in one year under a regime like a primary annual regime
        % the may specify a companion crop as well as the primary crop.
        function cropNames = getCropPlantedUnderRegimeInYear(regMgr, regimeLabel, year)
        
            % This method looks up the regime and then asks it the name of
            % the crops planted in the given year. There could be more than
            % one in the case of a primary regime that specifies a
            % companion crop in a given year.
            reg = regMgr.regimes(strcmp(regimeLabel, {regMgr.regimes.regimeLabel}));
            if isempty(reg)
                cropNames = {};
            else
                cropNames = reg.getCropsPlantedInYear(year);
            end
        end
        
        
        % This method returns a paddockLayout containing the information required to
        % correctly render the layout for the paddock in the requested
        % year.
        % The paddockLayout has information on the foreground colour and
        % the background colour, as well as whether borders, belts,
        % woodlands and contours are to be drawn. This information
        % ultimately comes from the regimes that are defined over the given
        % year, but is combined by the regimeManager into the paddockLayout
        % object, ultimately for the ImagineWindowManager to use.
        function pl = getPaddockLayoutInYear(regMgr, year)
        
            % This method finds all the regimes that are defined for the 
            % supplied year. It requests the paddockLayouts for those
            % years and combines them into a single paddockLayout which is
            % returned.
            
            % Find all the regimes that cover the supplied year.
            regs = regMgr.regimes([regMgr.regimes.startYear] >= year & [regMgr.regimes.finalYear] <= year);

            if isempty(reg)
                pl = PaddockLayout; % A new, empty PaddockLayout. All properties at default values.
            else
                pl = PaddockLayout; % Start with a new, empty PaddockLayout
                for i = 1:length(regs)
                    pl = pl.mergeWithPaddockLayout(regs(i).getPaddockLayoutInYear(year));                   
                end
                % At the end of the above for loop all regimes covering the supplied year will
                % have had input into the returned PaddockLayout.
            end
        end
        
        % This method creates and returns a PaddockSummary object that
        % contains the information required to know what is being planted
        % under which regimes in a given year. Also contains layout
        % information to assist with drawing the layout for the given year
        % in the main Imagine window.
        % 
        % The RegimeManager can be thought of as a factory for
        % PaddockSummary objects. It maintains a list of seperate regimes,
        % but is able to combine the information from those into
        % PaddockSummarys for each year.
        %
        % The two methods above, getCropsPlantedInYear(regimeLabel, year)
        % and getPaddockLayoutInYear(year) are superceded by this method, 
        % which is meant to be the sole interface to other objects. The two
        % other methods might be removed later.
        function ps = getPaddockSummaryForYear(regMgr, year)
           
            % Create the new, empty PaddockSummary
            ps = PaddockSummary();
            ps.year = year;
            ps.paddockLayout = PaddockLayout;
                        
            % Find all the regimes that cover the supplied year.
            regs = regMgr.regimes([regMgr.regimes.startYear] <= year & [regMgr.regimes.finalYear] >= year);
           
            % Hopefully, there are at most 2 regimes returned in regs; the
            % primary and secondary regimes.
            primReg = regs(strcmp({regs.type}, 'primary'));
            if ~isempty(primReg)
               if length(primReg) > 1 
                  print(['Found more than one primary regime covering year ', year, '.']); 
               else
                  ps.primaryRegimeCategory = primReg.regimeCategory;
                  ps.primaryRegimeLabel = primReg.regimeLabel;
                  cropNames = primReg.getCropsPlantedInYear(year);
         
                  if ~isempty(cropNames)
                      ps.primaryCropName = cropNames{1};
                      if length(cropNames) > 1
                          ps.companionCropName = cropNames{2};
                      end
                  end   
                  % Merge the primary paddockLayout into the
                  % PaddockSummary's paddockLayout
            
                  ps.paddockLayout = ps.paddockLayout.mergeWithPaddockLayout(primReg.getPaddockLayoutInYear(year));     
            
               end
            end
            
            secondReg = regs(strcmp({regs.type}, 'secondary'));
            if ~isempty(secondReg)
               if length(secondReg) > 1 
                  print(['Found more than one primary regime covering year ', year, '.']); 
               else      
                  ps.secondaryRegimeCategory = secondReg.regimeCategory;
                  ps.secondaryRegimeLabel = secondReg.regimeLabel;
                  cropNames = secondReg.getCropsPlantedInYear(year);
                  if ~isempty(cropNames)
                      ps.secondaryCropName = cropNames{1};
                  end
                  % Merge the secondary paddockLayout into the
                  % PaddockSummary's paddockLayout
                  ps.paddockLayout = ps.paddockLayout.mergeWithPaddockLayout(secondReg.getPaddockLayoutInYear(year));
               end
            end
            
        end % end getPaddockSummaryForYear
        
        
        % requestRegimeInstallation returns an InstalledRegime if a regime
        % should start (be installed) in the monthIndex provided.
        % The InstalledRegime has a list of cropNames that exist in the
        % regime, the regimeLabel, and a list of plantingEvents.
        function inReg = requestRegimeInstallation(regMgr, zoneString, sim)
            
            inReg = InstalledRegime.empty(1, 0);
            
            % Go through the list of regimes to see if one matches the
            % zone and should start in the monthIndex provided.
            
            matchFound = false;
            
            for i = 1:length(regMgr.regimes)
                              
                % Check the start month
                if (regMgr.regimes(i).startYear - 1)* 12 + 1 == sim.monthIndex

                   % Check the zone
                   type = regMgr.regimes(i).type;   
                  
                   if strcmp(zoneString, 'primary') 
                       if strcmp(type, 'primary') || strcmp(type, 'exclusive')
                          % Then we've found a match
                          matchFound = true;
                       end
                   elseif strcmp(zoneString, 'secondary')
                       if strcmp(type, 'secondary')
                          % Then we've found a match
                          matchFound = true;
                       end
                   end
                   
                   % If we found a match construct the inReg and return it.
                   if matchFound
                      inReg = InstalledRegime(regMgr.regimes(i), sim, zoneString);
                      return
                  end
                    
                end
                
            end
            
        end
        
        % Returns the regime definitions for the regimes that have this
        % crop in them.
        function regDefs = regimesThatUseCrop(regMgr, cropName)
        
           regDefs = struct('regimeCategory', {}, 'type', {}, 'regimeLabel', {});
            
           for i = 1:length(regMgr.regimes)
               if any(strcmp(regMgr.regimes(i).cropNameList, cropName))
                   regDef.regimeCategory = regMgr.regimes(i).regimeCategory;
                   regDef.type = regMgr.regimes(i).type;
                   regDef.regimeLabel = regMgr.regimes(i).regimeLabel;
                   regDefs(end + 1) = regDef;  %#ok<AGROW>
               end
           end
            
        end
            
        
        % Checks that the regimeManager is capcable of supplying
        % sufficient data for a simulation.
        function TF = isReadyForSimulation(regMgr)
            
            TF = length(regMgr.regimes) >= 1;
            if ~TF
                return
            end
                
            % Checks that each regime is valid, that each crop
            % used is properly defined. This should be the case anyway.
            % Should check that no regimes overlap.
            for i = 1:length(regMgr.regimes)
                TF = Regime.isValid(regMgr.regimes(i));
                if ~TF
                    return
                end                
            end
            
            % Check the primary regimes don't overlap.
            primaryRegimes = regMgr.regimes(strcmp({regMgr.regimes.type}, 'primary'));
            if ~isempty(primaryRegimes)
                [~, sortIx] = sort([primaryRegimes.startYear]);
                primaryRegimes = primaryRegimes(sortIx);
                for i = 1:length(primaryRegimes)-1
                    % if the next start year is before the current final
                    % year we have an overlap.
                    if primaryRegimes(i+1).startYear <= primaryRegimes(i).finalYear
                        TF = false;
                        return
                    end
                end
            end
            
            % Check that secondary regimes don't overlap.
            secondaryRegimes = regMgr.regimes(strcmp({regMgr.regimes.type}, 'secondary'));
            if ~isempty(secondaryRegimes)
                [~, sortIx] = sort([secondaryRegimes.startYear]);
                secondaryRegimes = secondaryRegimes(sortIx);
                for i = 1:length(secondaryRegimes)-1
                    % if the next start year is before the current final
                    % year we have an overlap.
                    if secondaryRegimes(i+1).startYear <= secondaryRegimes(i).finalYear
                        TF = false;
                        return
                    end
                end
            end
            
        end
        
        % When a crop is removed from the list, we have to make sure that
        % we remove all the regimes that might use that crop.
        function cropWasRemoved(singleObj,src,evnt)
          % singleObj - the singleton instance of this class
          % src - object generating event
          % evnt - the event data
          % Remove each regime cited in evnt.regimesToRemove
          for i = 1:length(evnt.regimesToRemove)
            %  {singleObj.regimes.regimeLabel}
            %  evnt.regimesToRemove{i}
              ix = find(strcmp({singleObj.regimes.regimeLabel}, evnt.regimesToRemove{i}), 1, 'first');
              if ~isempty(ix)
                  r = singleObj.regimes(ix);
             %     rlab = r.regimeLabel;
             %     tf = evnt.forceRegimeRemoval;
                  singleObj.removeRegime(r.regimeLabel, evnt.forceRegimeRemoval);
              end
          end
        
        end
        
        function cropNameChanged(singleObj, src, evtData)
           
            % Notify all the regimeDelegates that a crop name has changed.
            for i = 1:length(singleObj.regimes)
                singleObj.regimes(i).cropNameWasChanged(evtData.previousName, evtData.newName);
            end
        end
        
    end % end methods
    
    
    % Private helper functions
    methods (Access = private)
    
        % Sorts the list of regimes by their label.
        function sortRegimes(regMgr)
    
            % Clear empty regimes if they exist, which they shouldn't
           regLabels = {regMgr.regimes.regimeLabel};
           logs = logical(ones(1, length(regLabels)));
           for i = 1:length(regLabels)
              if isempty(regLabels{i})
                 logs(i) = false;
              else
                  logs(i) = true;
              end
           end
           
           regMgr.regimes = regMgr.regimes(logs);
            
            if ~isempty(regMgr.regimes)
                regLabels = {regMgr.regimes.regimeLabel};
                [~, indices] = sort(regLabels);
                regMgr.regimes = regMgr.regimes(indices);
                disp(['regimes sorted?', num2str(issorted({regMgr.regimes.regimeLabel}))])
            end            
        end
    
    end
end

File: imagine/regime/regime_manager.py
from dotwiz import DotWiz
from pyjson5 import pyjson5

from imagine.regime import Regime
from imagine.regime.installed_regime import InstalledRegime
from imagine.regime.regime_registry import *
from imagine.util.load_series_data import resolve_series


# from RegimeListChangedEventData import \
#     RegimeListChangedEventData  # Assuming you have a Python class for RegimeListChangedEventData
# from PaddockLayout import PaddockLayout  # Assuming you have a Python class for PaddockLayout
# from PaddockSummary import PaddockSummary  # Assuming you have a Python class for PaddockSummary


class RegimeManager:
    _instance = None

    def __init__(self):
        if 1 < 1.0001:
            raise RuntimeError('Call get_instance() instead')

        # Properties
        self.listeners = []

        # Implement a singleton class.
        self.regimes = []

    @classmethod
    def get_instance(cls, loaded_obj=None):
        if loaded_obj and isinstance(loaded_obj, RegimeManager) and cls._instance != loaded_obj:
            cls._instance = loaded_obj
            print('Set _instance to loaded_obj.')
        elif not cls._instance:
            cls._instance = cls.__new__(cls)
            cls._instance._regime_manager_constructor()
        return cls._instance

    def _regime_manager_constructor(self):
        # Properties
        self.listeners = []

        # Implement a singleton class.
        self.regimes = []

        self._regime_manager_init()

    def _regime_manager_init(self):
        # TODO: If listeners are necessary in pimagine, implement the listeners. Otherwise remove.
        # from pimagine.crop.crop_manager import CropManager
        # crop_mgr = CropManager.get_instance()
        # self.listeners.append(crop_mgr.add_listener('CropRemoved', self.crop_was_removed))
        # self.listeners.append(crop_mgr.add_listener('CropNameChanged', self.crop_name_changed))
        pass

    @staticmethod
    def get_regime_types(cls):
        RT1 = {'regimeCategory': 'Annual', 'regimeType': 'primary', 'regClass': 'AnnualRegime'}
        RT2 = {'regimeCategory': 'Belt', 'regimeType': 'secondary', 'regClass': 'BeltRegime'}
        return [RT1, RT2]

    @staticmethod
    def loadobj(obj):
        print('loading regime manager')
        RegimeManager._regime_manager_init(obj)
        return obj

    def add_regime(self, regime_cat):
        RTs = self.get_regime_types()
        RT = next((r for r in RTs if r['regimeCategory'] == regime_cat), None)
        if RT is None:
            raise ValueError('Tried to add a regime in nonexistent regime category.')
        r = Regime(RT['regClass'])
        r.setup_regime()
        if r.is_valid():
            self.regimes.append(r)
            self.sort_regimes()
            print('Added regime')
            # TODO: Implement listeners later if necessary. Or remove.
            # evt_data = RegimeListChangedEventData(self.regime_definitions(), '', r.regime_label)
            # self.notify('RegimeAdded', evt_data)
            # print('Notified: Added regime')

    def remove_regime(self, r, force_tf=False):
        if isinstance(r, int):
            regime_index = r
            if not 1 <= regime_index <= len(self.regimes):
                raise ValueError('Regime index passed is out of bounds.')
        elif isinstance(r, str):
            regime_index = next((i for i, reg in enumerate(self.regimes) if reg.regime_label == r), None)
            if regime_index is None:
                raise ValueError('Cannot find regime with matching regimeLabel to remove.')
        else:
            raise TypeError('Argument passed to removeRegime must be a string with a valid regimeLabel '
                            'of the index of the regime in the regimeList.')

        r = self.regimes.pop(regime_index)

        # if force_tf or input(
        #         f'Are you sure you want to remove the regime with label: {r.regime_label}? (Yes/No)').lower() == 'yes':
        #     rLabel = r.regime_label
        #     # TODO: Implement events and notifications if necessary for pimagine, or remove.
        #     # evt_data = RegimeListChangedEventData(self.regime_definitions(), rLabel, '')
        #     # self.notify('RegimeRemoved', evt_data)

    # def edit_regime(self, r):
    #     pass  # Implementation omitted for brevity

    def replace_regime(self, regime_label, new_regime):
        regime_index = next((i for i, regime in enumerate(self.regimes) if regime.regime_label == regime_label), None)
        if regime_index is not None:
            if new_regime.is_valid():
                self.regimes[regime_index] = new_regime
                self.sort_regimes()

                # TODO: Implement events and notifications if necessary for pimagine, or remove.
                # evt_data = RegimeListChangedEventData(self.regime_definitions, regime_label, new_regime.regime_label)
                # self.notify('RegimeEditted', evt_data)

    @property
    def regime_definitions(self):
        labels = [reg.regime_label for reg in self.regimes]
        category_names = [reg.regime_category for reg in self.regimes]
        types = [reg.type for reg in self.regimes]
        start_years = [reg.start_year for reg in self.regimes]
        final_years = [reg.final_year for reg in self.regimes]
        timeline_colours = [reg.timeline_colour for reg in self.regimes]
        crop_name_lists = [reg.crop_name_list for reg in self.regimes]

        return [
            {'regimeLabel': label, 'categoryName': category, 'type': reg_type, 'startYear': start, 'finalYear': final,
             'timelineColour': colour, 'cropNameList': crop_list}
            for label, category, reg_type, start, final, colour, crop_list in
            zip(labels, category_names, types, start_years, final_years, timeline_colours, crop_name_lists)]

    def get_regime_definition(self, regime_label):
        reg = next((reg for reg in self.regimes if reg.regime_label == regime_label), None)
        if reg:
            return {'regimeLabel': reg.regime_label, 'categoryName': reg.regime_category, 'type': reg.type,
                    'startYear': reg.start_year, 'finalYear': reg.final_year, 'timelineColour': reg.timeline_colour,
                    'cropNameList': reg.crop_name_list}
        else:
            return None

    def get_crop_planted_under_regime_in_year(self, regime_label, year):
        reg = next((reg for reg in self.regimes if reg.regime_label == regime_label), None)
        if reg:
            return reg.get_crops_planted_in_year(year)
        else:
            return []

    #### PaddockLayout and PaddockSummary not implemented in pimagine.
    # def get_paddock_layout_in_year(self, year):
    #     regs = [reg for reg in self.regimes if reg.start_year <= year <= reg.final_year]
    #     if regs:
    #         paddock_layout = PaddockLayout()
    #         for reg in regs:
    #             paddock_layout = paddock_layout.merge_with_paddock_layout(reg.get_paddock_layout_in_year(year))
    #         return paddock_layout
    #     else:
    #         return PaddockLayout()
    #
    # def get_paddock_summary_for_year(self, year):
    #     ps = PaddockSummary()
    #     ps.year = year
    #     ps.paddock_layout = PaddockLayout()
    #
    #     primary_reg = next(
    #         (reg for reg in self.regimes if reg.type == 'primary' and reg.start_year <= year <= reg.final_year), None)
    #     if primary_reg:
    #         ps.primary_regime_category = primary_reg.regime_category
    #         ps.primary_regime_label = primary_reg.regime_label
    #         crop_names = primary_reg.get_crops_planted_in_year(year)
    #         if crop_names:
    #             ps.primary_crop_name = crop_names[0]
    #             if len(crop_names) > 1:
    #                 ps.companion_crop_name = crop_names[1]
    #         ps.paddock_layout = ps.paddock_layout.merge_with_paddock_layout(
    #             primary_reg.get_paddock_layout_in_year(year))
    #
    #     secondary_reg = next(
    #         (reg for reg in self.regimes if reg.type == 'secondary' and reg.start_year <= year <= reg.final_year), None)
    #     if secondary_reg:
    #         ps.secondary_regime_category = secondary_reg.regime_category
    #         ps.secondary_regime_label = secondary_reg.regime_label
    #         crop_names = secondary_reg.get_crops_planted_in_year(year)
    #         if crop_names:
    #             ps.secondary_crop_name = crop_names[0]
    #         ps.paddock_layout = ps.paddock_layout.merge_with_paddock_layout(
    #             secondary_reg.get_paddock_layout_in_year(year))
    #
    #     return ps

    def request_regime_installation(self, zone_string, sim):
        in_reg = None
        for reg in self.regimes:
            # Check if the start month of the regime matches the simulation month index
            if (reg.start_year - 1) * 12 == sim.month_index:
                # Check the zone and regime type to determine if it should be installed
                if zone_string == 'primary' and reg.type in ['primary', 'exclusive']:
                    in_reg = InstalledRegime(reg, sim, zone_string)
                elif zone_string == 'secondary' and reg.type == 'secondary':
                    in_reg = InstalledRegime(reg, sim, zone_string)
        return in_reg

    def regimes_that_use_crop(self, crop_name):
        reg_defs = []
        for reg in self.regimes:
            if crop_name in reg.crop_name_list:
                reg_defs.append(
                    {'regimeCategory': reg.regime_category, 'type': reg.type, 'regimeLabel': reg.regime_label})
        return reg_defs

    def is_ready_for_simulation(self):
        # Check if there are at least 1 regime
        ready = len(self.regimes) >= 1
        if not ready:
            return False

        # Check that each regime is valid and no regime overlaps
        for reg in self.regimes:
            if not Regime.is_valid(reg):
                return False

        # Check that primary regimes don't overlap
        primary_regimes = [reg for reg in self.regimes if reg.type == 'primary']
        if primary_regimes:
            primary_regimes.sort(key=lambda x: x.start_year)
            for i in range(len(primary_regimes) - 1):
                if primary_regimes[i + 1].start_year <= primary_regimes[i].final_year:
                    return False

        # Check that secondary regimes don't overlap
        secondary_regimes = [reg for reg in self.regimes if reg.type == 'secondary']
        if secondary_regimes:
            secondary_regimes.sort(key=lambda x: x.start_year)
            for i in range(len(secondary_regimes) - 1):
                if secondary_regimes[i + 1].start_year <= secondary_regimes[i].final_year:
                    return False

        return True

    def crop_was_removed(self, src, evnt):
        # Remove each regime cited in evnt.regimes_to_remove
        for regime_label in evnt.regimes_to_remove:
            reg = next((r for r in self.regimes if r.regime_label == regime_label), None)
            if reg:
                self.remove_regime(reg.regime_label, evnt.force_regime_removal)

    def crop_name_changed(self, src, evt_data):
        # Notify all the regimeDelegates that a crop name has changed
        for reg in self.regimes:
            reg.crop_name_was_changed(evt_data.previous_name, evt_data.new_name)

    def sort_regimes(self):
        # Clear empty regimes if they exist, which they shouldn't
        self.regimes = [reg for reg in self.regimes if reg.regime_label]
        if self.regimes:
            self.regimes.sort(key=lambda x: x.regime_label)


    # # TODO: Implement listeners and notification if necessary in pimagine, or remove.
    # def notify(self, event, event_data):
    #     pass  # Implementation omitted for brevity

    def load_scenario(self, scenario_folder):

        regime_sub_folder = scenario_folder / 'Regimes'
        if not regime_sub_folder.is_dir():
            raise ValueError(f'Scenario regime folder does not exist: {str(regime_sub_folder)}')

        # For each regime load it into the regime mananger.from
        self.regimes = []
        for file_path in regime_sub_folder.iterdir():
            if file_path.is_file():
                if file_path.suffix == '.json5' and file_path.name[0] != '_':
                    with open(file_path, 'r') as file:

                        data = pyjson5.load(file)
                        data = resolve_series(data)
                        data = DotWiz(data)

                        # Load regime. Regime is abstract, so we can't instantiate it directly.
                        # Create an instance of the specific class.
                        reg = globals()[data.regime_class_name](data)
                        reg.create_events()
                        self.regimes.append(reg)

